pkg/innerring/internal/blockchain/blockchain.go

```diff
@@ -0,0 +1,398 @@
+// LevelDB configures Blockchain to use LevelDB located in given path.
+func LevelDB(path string) StorageConfig {
+	return StorageConfig{
+		typ:  dbconfig.LevelDB,
+		path: path,
+	}
+}
+
+// InMemory configures Blockchain to use volatile RAM storage.
+func InMemory() StorageConfig {
+	return StorageConfig{
+		typ: dbconfig.InMemoryDB,
+	}
+}
+
+// PingConfig configures P2P pinging mechanism.
+type PingConfig struct {
+	// Interval between pings. Must be positive.
+	Interval time.Duration
+
+	// Time period to wait for pong. Must be positive.
+	Timeout time.Duration
+}
+
+// P2PConfig configures communication over Neo P2P protocol.
+type P2PConfig struct {
+	// Specifies the minimum number of peers a node needs for normal operation. Must
+	// not be greater than math.MaxInt32.
+	MinPeers uint
+
+	// Specifies how many peers node should try to dial when connection counter
+	// drops below the MinPeers value. Must be in range [1, math.MaxInt32].
+	AttemptConnPeers uint
+
+	// Limits maximum number of peers dealing with the node. Must be in range
+	// [1, math.MaxInt32].
+	MaxPeers uint
+
+	// Pinging mechanism.
+	Ping PingConfig
+
+	// Maximum duration a single dial may take. Must be positive.
+	DialTimeout time.Duration
+
+	// Interval between protocol ticks with each connected peer. Must be positive.
+	ProtoTickInterval time.Duration
+}
+
+// Config configures Blockchain. All fields are required.
+type Config struct {
+	// Application level error listener. Blockchain writes any internal error to the
+	// channel. Error returns of functions (e.g. New) are not pushed. The channel
+	// should be regularly checked in order to prevent blocking. Any pop-up error
+	// does not allow the Blockchain to fully work, therefore, if it is detected,
+	// the blockchain should be stopped.
+	ErrorListener chan<- error
+
+	// NEO wallet of the node. The wallet is used by Consensus and Notary services.
+	// Corresponding private key be accessed via Blockchain.LocalKey.
+	Wallet config.Wallet
+
+	// Storage configuration. Must be set using one of constructors like BoltDB.
+	Storage StorageConfig
+
+	// Identifier of the Neo network.
+	NetworkMagic netmode.Magic
+
+	// Initial committee staff.
+	Committee keys.PublicKeys
+
+	// Time period (approximate) between two adjacent blocks. Must be positive.
+	BlockInterval time.Duration
+
+	// P2P settings.
+	P2P P2PConfig
+
+	// Length of the chain accessible to smart contracts. Must be positive.
+	TraceableChainLength uint32
+}
+
+// Options tunes Blockchain's default behavior. All fields are optional: zero
+// (omitted) values don't change the behavior. Launched Blockchain should be
+// stopped.
+type Options struct {
+	// Writer of the Blockchain's logs and internal errors. By default, Blockchain
+	// doesn't write logs.
+	Logger *zap.Logger
+
+	// Network addresses to listen Neo P2P on. Each element must be a valid TCP
+	// address in 'host:port' format. By default, Neo P2P is not served.
+	P2PListenAddresses []string
```

Comment:
Make it a part of `P2PConfig`?

Reply:
Here I have separated the required (`Config`) and optional (`Options`) settings even if they belong to the same section. Do you think it's better to keep a single configuration structure and explicitly prescribe the optional settings?

Reply:
Optional ones can have some per-setting defaults, so they can be omitted if not needed.

Reply:
Currently omitted `Config` field is incorrect, and empty `Options` field defaults to prescribed value. Transferring field `P2PListenAddresses` to `P2PConfig` will mean that it is mandatory and cannot be omitted. Otherwise I will have to merge `Config` with `Options`.

Reply:
Just have a `P2P` section (yes, mandatory) and have some defaults for unconfigured fields there. It's OK for people to specify some fields, but not specify the others. Some P2P config must be present and this setting is organically a part of it.